### ISSUE TITLE
Async I/O java-option flags are ignored

### ISSUE THREAD

**Issue by jacobrh91** [broadinstitute/gatk#4435]
Users are supposed to enable/disable async I/O options by passing java option flags to the gatk launch shell script:

`
./gatk-4.0.1.2/gatk --java-options "-Dsamjdk.use_async_io_read_samtools=true -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=true" <Tool and its flags>
`

It appears that no matter how these flags are passed, and even if one hard codes the variables in the actual shell script (change the defaults within the script), they are ignored.

**Three examples follow:**

(Irrelevant sections are omitted with ". . ." in their place for readability) 
(These examples use BaseRecalibrator as an example tool)

**Example 1: No Flags Passed (but -Xmx12G)** 
`
./gatk-4.0.1.2/gatk --java-options "-Xmx12G" <Tool specific commands>
`

```
Using GATK jar /projects/bioinformatics/TestingGATK4/gatk-4.0.1.2/gatk-package-4.0.1.2-local.jar                        
Running:                                                                                                                
    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=1 -Xmx12G -jar /projects/bioinformatics/TestingGATK4/gatk-4.0.1.2/gatk-package-4.0.1.2-local.jar BaseRecalibrator -I /projects/bioinformatics/TestingGATK4/hg38_GATK4/async_options_run/newOutputs/hg38_aligned.sorted.dedupped.bam --known-sites /projects/bioinformatics/DataPacks/human/NA12878_fullsize_bundle_GATK4_copy_2/dbsnp_hg38/dbsnp_146.hg38.vcf -O /projects/bioinformatics/TestingGATK4/hg38_GATK4/async_options_run/newOutputs/recal.table -R /projects/bioinformatics/DataPacks/human/NA12878_fullsize_bundle_GATK4_copy_2/reference_hg38/Homo_sapiens_assembly38.fasta
10:45:49.790 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/projects/bioinformatics/TestingGATK4/gatk-4.0.1.2/gatk-package-4.0.1.2-local.jar!/com/intel/gkl/native/libgkl_compression.so
10:45:49.887 INFO  BaseRecalibrator - ------------------------------------------------------------                      
10:45:49.888 INFO  BaseRecalibrator - The Genome Analysis Toolkit (GATK) v4.0.1.2                                       
. . .
10:45:49.889 INFO  BaseRecalibrator - HTSJDK Version: 2.14.1                                                            
10:45:49.889 INFO  BaseRecalibrator - Picard Version: 2.17.2                                                            
10:45:49.889 INFO  BaseRecalibrator - HTSJDK Defaults.COMPRESSION_LEVEL : 1                                             
10:45:49.889 INFO  BaseRecalibrator - HTSJDK Defaults.USE_ASYNC_IO_READ_FOR_SAMTOOLS : false                            
10:45:49.889 INFO  BaseRecalibrator - HTSJDK Defaults.USE_ASYNC_IO_WRITE_FOR_SAMTOOLS : true                            
10:45:49.889 INFO  BaseRecalibrator - HTSJDK Defaults.USE_ASYNC_IO_WRITE_FOR_TRIBBLE : false 
```

**Result:** No flags were passed, so the default values are passed to java at runtime (seen in the second line)

**Example 2: Flags Passed (Along with -Xmx12G)**
`
./gatk-4.0.1.2/gatk --java-options "-Xmx12G -Dsamjdk.use_async_io_read_samtools=true -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=true" <Tool specific commands>
`

```
Using GATK jar /projects/bioinformatics/TestingGATK4/gatk-4.0.1.2/gatk-package-4.0.1.2-local.jar                        
Running:                                                                                                                
    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=1 -Xmx12G -Dsamjdk.use_async_io_read_samtools=true -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=true -jar /projects/bioinformatics/TestingGATK4/gatk-4.0.1.2/gatk-package-4.0.1.2-local.jar BaseRecalibrator -I /projects/bioinformatics/TestingGATK4/hg38_GATK4/async_options_run/newOutputs/hg38_aligned.sorted.dedupped.bam --known-sites /projects/bioinformatics/DataPacks/human/NA12878_fullsize_bundle_GATK4_copy_2/dbsnp_hg38/dbsnp_146.hg38.vcf -O /projects/bioinformatics/TestingGATK4/hg38_GATK4/async_options_run/newOutputs/recal.table -R /projects/bioinformatics/DataPacks/human/NA12878_fullsize_bundle_GATK4_copy_2/reference_hg38/Homo_sapiens_assembly38.fasta
11:13:22.175 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/projects/bioinformatics/TestingGATK4/gatk-4.0.1.2/gatk-package-4.0.1.2-local.jar!/com/intel/gkl/native/libgkl_compression.so
11:13:22.277 INFO  BaseRecalibrator - ------------------------------------------------------------                      
11:13:22.277 INFO  BaseRecalibrator - The Genome Analysis Toolkit (GATK) v4.0.1.2                                       
. . .
11:13:22.278 INFO  BaseRecalibrator - HTSJDK Version: 2.14.1                                                            
11:13:22.278 INFO  BaseRecalibrator - Picard Version: 2.17.2                                                            
11:13:22.278 INFO  BaseRecalibrator - HTSJDK Defaults.COMPRESSION_LEVEL : 1                                             
11:13:22.278 INFO  BaseRecalibrator - HTSJDK Defaults.USE_ASYNC_IO_READ_FOR_SAMTOOLS : false                            
11:13:22.278 INFO  BaseRecalibrator - HTSJDK Defaults.USE_ASYNC_IO_WRITE_FOR_SAMTOOLS : true                            
11:13:22.278 INFO  BaseRecalibrator - HTSJDK Defaults.USE_ASYNC_IO_WRITE_FOR_TRIBBLE : false
```
**Result:** The flags are passed to java, but only after the default flags are also passed. Regardless, the actual behavior of the tool is unchanged (as seen in the last three lines)

**Examples 3: Hardcoded default values in shell script were edited**

```
cat ./gatk-4.0.1.2/gatk
. . .
PACKAGED_LOCAL_JAR_OPTIONS= ["-Dsamjdk.use_async_io_read_samtools=true",                                               
                  "-Dsamjdk.use_async_io_write_samtools=true",                                                          
                  "-Dsamjdk.use_async_io_write_tribble=true",                                                          
                  "-Dsamjdk.compression_level=1"]
. . .
```
`
./gatk-4.0.1.2/gatk --java-options "-Xmx12G" <Tool specific commands>
`
```
Using GATK jar /projects/bioinformatics/TestingGATK4/gatk-4.0.1.2/gatk-package-4.0.1.2-local.jar
Running:
    java -Dsamjdk.use_async_io_read_samtools=true -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=true -Dsamjdk.compression_level=1 -Xmx12G -jar /projects/bioinformatics/TestingGATK4/gatk-4.0.1.2/gatk-package-4.0.1.2-local.jar BaseRecalibrator -I /projects/bioinformatics/TestingGATK4/hg38_GATK4/async_options_run/newOutputs/hg38_aligned.sorted.dedupped.bam --known-sites /projects/bioinformatics/DataPacks/human/NA12878_fullsize_bundle_GATK4_copy_2/dbsnp_hg38/dbsnp_146.hg38.vcf -O /projects/bioinformatics/TestingGATK4/hg38_GATK4/async_options_run/newOutputs/recal.table -R /projects/bioinformatics/DataPacks/human/NA12878_fullsize_bundle_GATK4_copy_2/reference_hg38/Homo_sapiens_assembly38.fasta
11:23:11.500 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/projects/bioinformatics/TestingGATK4/gatk-4.0.1.2/gatk-package-4.0.1.2-local.jar!/com/intel/gkl/native/libgkl_compression.so
11:23:11.637 INFO  BaseRecalibrator - ------------------------------------------------------------
11:23:11.637 INFO  BaseRecalibrator - The Genome Analysis Toolkit (GATK) v4.0.1.2
. . .
11:23:11.638 INFO  BaseRecalibrator - HTSJDK Version: 2.14.1
11:23:11.639 INFO  BaseRecalibrator - Picard Version: 2.17.2
11:23:11.639 INFO  BaseRecalibrator - HTSJDK Defaults.COMPRESSION_LEVEL : 1
11:23:11.639 INFO  BaseRecalibrator - HTSJDK Defaults.USE_ASYNC_IO_READ_FOR_SAMTOOLS : false
11:23:11.639 INFO  BaseRecalibrator - HTSJDK Defaults.USE_ASYNC_IO_WRITE_FOR_SAMTOOLS : true
11:23:11.639 INFO  BaseRecalibrator - HTSJDK Defaults.USE_ASYNC_IO_WRITE_FOR_TRIBBLE : false
```
**Result:** Now the changed values show up and are passed to Java, but the tool still runs in the same way

**Comment by droazen** [comment]
@jacobrh91 Thanks for the report! I believe this is the result of our (still-incomplete) transition to using a GATK configuration file for certain toolkit-wide settings. Could you please try making a copy of `src/main/resources/org/broadinstitute/hellbender/utils/config/GATKConfig.properties`, editing the `samjdk.use_async_io_read_samtools` line in that file to have a value of `true` rather than `false`, pass in the edited config file to GATK via the `--gatk-config-file` argument, and see if the setting gets changed at runtime?

**Comment by droazen** [comment]
@jonn-smith Can you please chime in here as well to tell us whether it's currently possible to override a single system property setting in Owner on the command line, or is `--gatk-config-file` currently the only way?

**Comment by jonn-smith** [comment]
Currently the code for the config file parsing will override values passed in on the command-line, so that explains one reason this is happening.

The fix should be pretty straightforward.

Opened #4436

**Comment by jacobrh91** [comment]
Passing in the properties file did work (with the --gatk-config-file option)

However, of the four tools I tested (MarkDuplicates, BaseRecalibrator, ApplyBQSR, and HaplotypeCaller) all of the tools accepted the --gatk-config-file option except for MarkDuplicates, which complains that it is not a recognized option.

Perhaps this should be turned into a separate issue?

Thanks

**Comment by jonn-smith** [comment]
@jacobrh91 Yeah - can you submit that as a separate issue?

**Comment by cmnbroad** [comment]
Unlike the other 3 tools, MarkDuplicates isn't a native GATK tool, its a Picard tool, so it won't recognize the standard GATK options. Not sure if we can/should to try to do something to minimize the confusion around that.
